### PR TITLE
Self-extracting: add --exec param, allow compressed to be renamed

### PR DESCRIPTION
**Pull request by yakov-olkhovskiy**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

New parameter --exec for compressor allows to explicitly specify executable to be executed upon decompression. Compressed executable can have different name than one selected for execution, and running executable will replace compressed under its name - this also allows for compressed executable to be renamed.

### CODE REVIEW

File: utils/self-extracting-executable/decompressor.cpp [hunk 1]
```diff
@@ -226,8 +226,7 @@ int decompressFiles(int input_fd, char * path, char * name, bool & have_compress
         file_info = *reinterpret_cast<FileData*>(input + files_pointer);
         files_pointer += sizeof(FileData);
 
-        size_t file_name_len =
-            (strcmp(input + files_pointer, name) ? le64toh(file_info.name_length) : le64toh(file_info.name_length) + 13 + 7);
+        size_t file_name_len = file_info.exec ? strlen(name) + 13 + 7 + 1 : le64toh(file_info.name_length);
```
Comment (vdimir) [inline utils/self-extracting-executable/decompressor.cpp:229]: What are 13 and 7 for?

File: utils/self-extracting-executable/compressor.cpp [hunk 2]
```diff
@@ -253,14 +279,16 @@ int compressFiles(char* filenames[], int count, int output_fd, int level, const
...
+        names[i] = strrchr(filename, '/');
         if (names[i])
             ++names[i];
         else
-            names[i] = filenames[i];
+            names[i] = filename;
         size_t nlen = strlen(names[i]) + 1;
         files_data[i].name_length = htole64(nlen);
         sum_file_size += nlen;
+        if (!is_exec && !exec && strcmp(names[i], out_name) == 0)
```
Comment (vdimir) [inline utils/self-extracting-executable/compressor.cpp:290]:
```suggestion
        /// if no `--exec` is provided, use the out filename as executable
        if (!is_exec && !exec && strcmp(names[i], out_name) == 0)
```

File: utils/self-extracting-executable/compressor.cpp [hunk 3]
```diff
@@ -228,21 +239,36 @@ int saveMetaData(char* filenames[], int count, int output_fd, const MetaData& me
...
     {
-        printf("Compressing: %s\n", filenames[i]);
+        const char* filename = nullptr;
+        if (i == count)
+        {
+            if (!is_exec)
+                continue;
+            filename = exec;
+            files_data[i].exec = true;
+        }
```
Comment (vdimir) [inline utils/self-extracting-executable/compressor.cpp:265]: Consider extracting the loop body into function and make one explicit extra call.

**Review by vdimir — APPROVED** [review]

File: utils/self-extracting-executable/compressor.cpp (hunk 3 above)
Comment (yakov-olkhovskiy) [reply utils/self-extracting-executable/compressor.cpp:265]: this will involve too many arguments making it even more obscure
let's leave it as is...

**Pull request merged**